moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp

```diff
@@ -338,31 +373,29 @@ class MoveItControllerManager : public moveit_controller_manager::MoveItControll
       ControllersMap::iterator c = managed_controllers_.find(it);
       if (c != managed_controllers_.end())
       {  // controller belongs to this manager
-        srv.request.start_controllers.push_back(c->second.name);
-        for (controller_manager_msgs::HardwareInterfaceResources& claimed_resource : c->second.claimed_resources)
+        request->start_controllers.push_back(c->second.name);
+        for (const auto& claimed_resource : c->second.claimed_interfaces)
         {
-          for (const std::string& resource : claimed_resource.resources)
-          {  // for all claimed resource
-            resources_bimap::right_const_iterator res = claimed_resources.right.find(resource);
-            if (res != claimed_resources.right.end())
-            {                                                       // resource is claimed
-              srv.request.stop_controllers.push_back(res->second);  // add claiming controller to stop list
-              claimed_resources.left.erase(res->second);            // remove claimed resources
-            }
+          resources_bimap::right_const_iterator res = claimed_resources.right.find(claimed_resource);
+          if (res != claimed_resources.right.end())
+          {                                                    // resource is claimed
+            request->stop_controllers.push_back(res->second);  // add claiming controller to stop list
+            claimed_resources.left.erase(res->second);         // remove claimed resources
           }
         }
       }
     }
-    srv.request.strictness = srv.request.STRICT;
+    request->strictness = controller_manager_msgs::srv::SwitchController::Request::STRICT;
 
-    if (!srv.request.start_controllers.empty() || srv.request.stop_controllers.empty())
+    if (!request->start_controllers.empty() || request->stop_controllers.empty())
     {  // something to switch?
-      if (!ros::service::call(getAbsName("controller_manager/switch_controller"), srv))
+      auto result_future = switch_controller_service_->async_send_request(request);
+      if (result_future.wait_for(std::chrono::duration<double>(SERVICE_CALL_TIMEOUT)) == std::future_status::timeout)
       {
-        ROS_ERROR_STREAM("Could switch controllers at " << ns_);
+        RCLCPP_ERROR_STREAM(LOGGER, "Could switch controllers at " << switch_controller_service_->get_service_name());
       }
       discover(true);
-      return srv.response.ok;
+      return result_future.get()->ok;
```

Comment:
In case of a timeout `get()` shouldn't be called anymore since it's using an internal wait call.